### ISSUE TITLE
constructor call not found for member object

### ISSUE THREAD

**Issue by Jacob-Burckhardt** [clangd/clangd#1312]
### Observed behavior

After asking for calls to Blob() it finds a call on line 8 but not on line 7.

```
class Blob {
public:
   Blob() {} // line 3
};

class Chunk {
   Blob b; // line 7
   Blob *p = new Blob; // line 8
};

int main() {
   Chunk c;
}
```

### Steps to reproduce

1. Put the cursor on "B" in the Blob declaration on line 3.
2. In emacs, run M-x lsp-find-references to find calls to the Blob constructor.

**Logs**

```
V[17:01:37.358] <<< {"id":84,"jsonrpc":"2.0","method":"textDocument/references","params":{"context":{"includeDeclaration":false},"position":{"character":4,"line":2},"textDocument":{"uri":"file:///home/bjacob/backup/C%2B%2B/tests/a.cpp"}}}

I[17:01:37.358] <-- textDocument/references(84)
V[17:01:37.358] ASTWorker running References on version 14 of /home/bjacob/backup/C++/tests/a.cpp
I[17:01:37.359] --> reply:textDocument/references(84) 0 ms
V[17:01:37.359] >>> {"id":84,"jsonrpc":"2.0","result":[{"range":{"end":{"character":21,"line":7},"start":{"character":17,"line":7}},"uri":"file:///home/bjacob/backup/C%2B%2B/tests/a.cpp"}]}
```

**System information**

Output of `clangd --version`:
```
clangd version 15.0.1 (https://github.com/llvm/llvm-project b73d2c8c720a8c8e6e73b11be4e27afa6cb75bdf)
Features: linux+grpc
Platform: x86_64-unknown-linux-gnu
```

Editor/LSP plugin: Emacs 27.1 / lsp-mode-7.0.1

Operating system: Debian GNU/Linux 11

**Comment by HighCommander4** [comment]
What source location would you expect to be recorded for this reference - the field name?

What if `C` had a constructor which initialized `b` using a different constructor in a ctor-initializer -- would you expect the reference not to be recorded in that case?

**Comment by Jacob-Burckhardt** [comment]
> What source location would you expect to be recorded for this reference - the field name?

I think the field name is a good choice.  The class name would be ok too.  Note that in the below example it finds the constructor call on line 7 and puts my cursor on the variable name "b" which I think is good behavior:

```
class Blob {
public:
   Blob() {} // line 3
};

int main() {
   Blob b; // line 7
}
```

> 
> What if `C` had a constructor which initialized `b` using a different constructor in a ctor-initializer -- would you expect the reference not to be recorded in that case?

I would still expect it to be recorded and I am glad to see that clangd does that.  Below it puts the cursor on "b" on line 8:

```
class Blob {
public:
   Blob(int) {} // line 3
};

class Chunk {
public:
   Chunk() : b{9} {} // line 8
   
   Blob b;
};

int main() {
   Chunk c;
}
```

**Comment by HighCommander4** [comment]
Sorry, let me clarify my last question. In a situation like this:

```c++
class Blob {
public:
  Blob() {}
  Blob(int) {}
};

class Chunk {
public:
  Chunk() : b(9) {}

  Blob b;
};
```

should the field declaration no longer reference the *default constructor* of `Blob`?

What about if `Chunk` has two constructors, one of which calls the default constructor of `Blob`:

```c++
class Chunk {
public:
  Chunk() {}
  Chunk() : b(9) {}

  Blob b;
};
```

**Comment by HighCommander4** [comment]
(What I'm getting at here is that whether or not the `Blob` default constructor is called depends on how the `Chunk` constructors are implemented, making the field declaration a somewhat unintuitive location for a reference to it.)

**Comment by Jacob-Burckhardt** [comment]
I will try to respond by using this more complete example:

```
#include <cassert>

class Blob {
public:
   Blob() { assert(0); }
   Blob(int) { assert(0);  }
};

class Chunk {
public:
  Chunk() {}
  Chunk(int) : b(9) {}

  Blob b;
};

int main() {
   Chunk c;
}
```

For the second Chunk ctor, I would like the reference to the Blob ctor to be on the Chunk ctor definition.  For consistency, the first Chunk ctor definition could also be referenced.  However, I realize that might be confusing to some users since they asked for Blob ctor references but yet they get back a line which shows no Blob types and no variables of type Blob.  It will be even more confusing if Chunk has a hundred methods and fields and the user is unaware that Chunk contains Blob.  

In the above example, gdb prints a call stack for the assertion failure like this:

```
#3  0x00007ffff7ae9662 in __GI___assert_fail (assertion=0x402004 "0", file=0x402006 "a.cpp", line=5,
    function=0x40200c "Blob::Blob()") at assert.c:101
#4  0x00000000004011a4 in Blob::Blob (this=0x7fffffffe2d8) at a.cpp:5
#5  0x0000000000401165 in Chunk::Chunk (this=0x7fffffffe2d8) at a.cpp:11
#6  0x0000000000401141 in main () at a.cpp:18
```
gdb created the same confusion by saying that the default Chuck ctor calls the Blob ctor even though there is no explicit call in the referenced code line 11.  If people are used to debuggers doing this then hopefully they can get used to clangd doing it.  

Maybe clangd could avoid that confusion by making the editor display the text "implicit call to default Blob ctor".  But I realize that clangd is the server and clangd maintainers might not have enough control over editors to make them display that message.

If you comment out both Chunk ctors, then gdb says that Chunk::Chunk calls Blob::Blob which can also be confusing since there is no explicit Chunk::Chunk in the source code.  In such a backtrace, gdb says the Chunk::Chunk is on the line that says "class Chunk".  So I think it would be reasonable for clangd to say that a call exists there as well.

On the other hand, if clangd showed a reference on the Blob field declaration and another reference on the second Chunk ctor, definition then it would be inconsistent, but maybe that is less confusing to users.

I think any of the above techniques would be an improvement to clangd and I can't decide which technique I like best.

However, I do feel strongly that if a ctor call is in an initializer list then the reference should be in that list so that I can immediately see the arguments that are passed to Blob ctor.

**Comment by Jacob-Burckhardt** [comment]
I wrote:

> If you comment out both Chunk ctors, then gdb says that Chunk::Chunk calls Blob::Blob which can also be confusing since there is no explicit Chunk::Chunk in the source code. In such a backtrace, gdb says the Chunk::Chunk is on the line that says "class Chunk". So I think it would be reasonable for clangd to say that a call exists there as well.

I changed my mind after I thought of another example.  Putting the reference on "class Chunk" would be bad for the below example since that means that both b1 and b2 would be referenced on "class Chunk".  Both references would look identical which would confuse users and users might not realize that there are _two_ continued Blobs.  But if you put the reference on the field declaration then the references will differ in that each will cover a different field name which will make it clear there are two Blobs.

So to summarize, I think that if a field is mentioned in an initializer list the reference should be to that list.  If an explicit or implicit ctor does not refer to a field, then the reference should be the field declaration.

```
class Blob {
public:
   Blob() {} // line 3
};

class Chunk {
   Blob b1, b2; // line 7
};

int main() {
   Chunk c;
}
```
